### PR TITLE
Popover: fix arrow placement and design

### LINKED ISSUE #42820: Popover arrow not working properly in the latest versions
### Description

Popover arrow is working well on WP 6.0, but when I activate the Gutenberg plugin (from the repository), it's not working properly anymore.

### Step-by-step reproduction instructions

The way to reproduce is through code.

1. In the editor, add a popover with the prop `noArrow={true}`.
2. Open it in an env with WP 6.0 and see it working properly.
3. Open it in an env with WP 6.0, and Gutenberg plugin ([trunk](https://github.com/WordPress/gutenberg/tree/ed5813c1325b8d6a09df7dd1bda00a8d90faf820)) activated, and see that it doesn't work properly anymore.

### Screenshots, screen recording, code snippet

Using WP 6.0 with Gutenberg plugin deactivated (working):

With Gutenberg ([trunk](https://github.com/WordPress/gutenberg/tree/ed5813c1325b8d6a09df7dd1bda00a8d90faf820)) plugin activated (not working):

A case with Gutenberg plugin activated, with the Popover opening in the bottom (the arrow appears, but in a weird way):

Screen Shot 2022-07-29 at 10 46 23

A code snippet to test:

```js
<Popover noArrow={ false }>
  <div style={ { width: '400px', height: '400px' } }>
    Lorem ipsum
  </div>
</Popover>
```

### Environment info

- WordPress 6.0 with Gutenberg plugin activated ([trunk](https://github.com/WordPress/gutenberg/tree/ed5813c1325b8d6a09df7dd1bda00a8d90faf820)) in Chrome.

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by ciampo**
Fixes #42820

## What?

Fix a regression introduced in #40740 about the `Popover`'s arrow placement.

## Why?

The `Popover`'s arrow has currently a few issues:

- it is not always positioned correctly
- it doesn't render with the same background color and border as the popover's contents
- its size is smaller than it used to be before #40740

## How?

In order to fix the arrow placement, the following changes were implemented:

- fixed a bug in the logic behind validating the arrow's `x` and `y` coordinates received from `useFloating` (mainly switching from `! Number.isNaN` to `Number.isFinite`)
- changed the approach in rendering the arrow's triangle:
  - from a diamond rotated `45deg`, switched to a proper triangle (coded in SVG)
  - Added (back) support for a border color around the triangle, in line with how the design used to be
  - moved some inline styles to SASS styles, using an internal `$arrow-triangle-base-size` SCSS variable

I've also added two extra temporary commits (which I will remove before merging)
  - one adding a Storybook example to easily check that the arrow's placement is correct for all possible placements
  - another one adding a quick change that makes sure that `ownerDocument` is always defined inside `Popover` (this fix should probably be moved to its own separate PR)

## Testing Instructions

- In Storybook: on your machine, visit the new temporary "Positioning" Storybook example and make sure that the arrow's placement and design looks correct for all possible popover placements
- In the WordPress app: on your machine, visit `/wp-admin/site-editor.php?postType=wp_template` and click the "Add New" button in the top right part of the screen. Make sure that the Popover's arrow placement and design look correct (see [this comment](https://github.com/WordPress/gutenberg/issues/42820#issuecomment-1200612126) for more details)
- Try to replicate the bug reported in [the original issue's description](https://github.com/WordPress/gutenberg/issues/42820#issue-1322276575) and confirm that it can't be replicated anymore

For extra thoroughness, you could revert the `Popover: fix arrow logic, use SVG to render triangle` commit in this PR and verify that the bug is not fixed in the scenarios described above.

## Screenshots or screencast 

| Before | After |
|---|---|
|  image | Screenshot 2022-08-01 at 23 19 57 |

**Before:**

https://user-images.githubusercontent.com/1083581/182247869-27cb9797-e8d3-4cd8-a45e-d43d89f52774.mp4

**After:**

https://user-images.githubusercontent.com/1083581/182247281-7f07e868-6305-40b2-9a8c-40b4f0bb2e0b.mp4

### CODE REVIEW

File: packages/components/src/popover/index.js [hunk 1]
```diff
@@ -407,22 +428,25 @@ const Popover = (
 			<div className="components-popover__content">{ children }</div>
 			{ hasArrow && (
 				<div
-					className="components-popover__arrow"
 					ref={ arrowRef }
+					className={ [
+						'components-popover__arrow',
+						`components-popover__arrow--${
+							placementData.split( '-' )[ 0 ]
+						}`,
```
Comment (mirka) [inline packages/components/src/popover/index.js]: The [CSS coding guidelines](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/coding-guidelines.md#css) in this repo are "inspired by BEM" but not exactly BEM, so we want the modifier classes to be named `.is-top`, `.is-right`, etc.

File: packages/components/src/popover/style.scss [hunk 2]
```diff
@@ -60,12 +60,64 @@
...
+	width: 100%;
+	stroke-width: $border-width;
+
+	// Same as popover content
+	&-bg {
+		fill: $white;
+		stroke: $white;
+	}
+
+	&-border {
```
Comment (mirka) [inline packages/components/src/popover/style.scss]: Nit: This might be a personal preference thing, but I prefer full class names to be written out (`.components-popover__triangle-bg`) so they can be searched easily 🥺

File: packages/components/src/popover/style.scss [hunk 3]
```diff
@@ -60,12 +60,64 @@
...
 	position: absolute;
-	background: $gray-400;
-	width: 8px;
-	height: 8px;
-	transform: rotate(45deg);
-	z-index: -1;
-	.is-alternate & {
-		background: $gray-900;
+	width: var(--wp-popover-arrow-size);
+	height: var(--wp-popover-arrow-size);
```
Comment (mirka) [inline packages/components/src/popover/style.scss]: Nice 💛

**Review by mirka — APPROVED** [review]
Looks great, thanks for the quick fix!

At first I was wondering why the Add New Template dropdown is offset a little bit lower than [before](https://user-images.githubusercontent.com/555336/182060058-c240c365-1f40-4165-8e83-001a5069d410.png) and no longer overlaps the header, but I guess we can consider this a consistency improvement since it's the same offset as other Dropdowns without the arrow. Like the Options dropdown in the editor:

The Options dropdown in the editor sits flush with the header

File: packages/components/src/popover/index.js [hunk 4]
```diff
@@ -137,25 +162,26 @@ const Popover = (
...
-			return (
+			documentToReturn =
 				getAnchorRect( anchorRefFallback.current )?.ownerDocument ??
-				document
-			);
+				document;
 		}
 
-		return document;
+		return documentToReturn ?? document;
```
Comment (ntsekouras) [inline packages/components/src/popover/index.js]: I guess there are cases that `documentToReturn` is not defined to make this change.. Can you explain where that is and if you know why? Do we miss something in our current handling of `anchorRef`?

**Comment by ciampo** [comment]
> At first I was wondering why the Add New Template dropdown is offset a little bit lower than [before](https://user-images.githubusercontent.com/555336/182060058-c240c365-1f40-4165-8e83-001a5069d410.png) and no longer overlaps the header, but I guess we can consider this a consistency improvement since it's the same offset as other Dropdowns without the arrow.

I see both "sides" of the argument, but I agree with your point — this could be seen as a step towards consistency. But even if we changed our mind and wanted to move the popover closer to the "Add new" button, I would suggest that we do it in a separate PR so that we keep this one focused on the arrow fix.

File: packages/components/src/popover/index.js [hunk 5]
```diff
@@ -407,22 +428,25 @@ const Popover = (
...
-						right: undefined,
-						bottom: undefined,
-						[ staticSide ]: '-4px',
+						left: Number.isFinite( arrowData?.x )
+							? `${ arrowData.x }px`
+							: '',
+						top: Number.isFinite( arrowData?.y )
+							? `${ arrowData.y }px`
+							: '',
+						'--wp-popover-arrow-size': '14px',
```
Comment (ntsekouras) [inline packages/components/src/popover/index.js]: So `14px` is our magic number and I guess it's the one to match our previous design, correct? 

Since this number is not affected by anything and is not configurable, is there a reason to have it here and not in the scss file directly?

File: packages/components/src/popover/index.js (hunk 1 above)
Comment (ntsekouras) [reply packages/components/src/popover/index.js]: Super nit: why not just use string interpolation here?

**Review by ntsekouras — APPROVED** [review]
Awesome work here Marco! 💯 

I took a brief look at this yesterday and saw the many things missing from arrow handling. I've tested your PR by changing every Dropdown we have to show an arrow and everything seemed great(both in post and site editors).

I've left a few small comments/questions and I'd like to know more especially for [this one](https://github.com/WordPress/gutenberg/pull/42874/files#r935191862). I know you'll have some small cleanup to do, but this looks good to me!

**Comment by talldan** [comment]
FWIW, the Add New Template dropdown probably shouldn't have an arrow. I think it's the only dropdown in the codebase that has one, so that could be seen as an inconsistency.

File: packages/components/src/popover/index.js (hunk 4 above)
Comment (ciampo) [reply packages/components/src/popover/index.js]: Sure! I had to add this change because, when I created the temporary Storybook example, I decided to [pass the anchor to the popover via `anchorRef`](https://github.com/WordPress/gutenberg/pull/42874/commits/759d0348398f44c820ab8d50d15275cca1b2ab6b#diff-88eb39a8c17230c5315615af25bf97fe202a13447ab24d5c6cf3cf292ff6ba82R111).

That caused an error because `ownerDocument` was `undefined` — and so I inspected the code:

https://github.com/WordPress/gutenberg/blob/759d0348398f44c820ab8d50d15275cca1b2ab6b/packages/components/src/popover/index.js#L139-L159

Basically, what I believe was happening was:
- `anchorRef` is a "modern style" ref, and therefore it's an object with the `current` prop
- initially, `anchorRef.current` is `null`
- therefore, the check on line `144` (`if ( anchorRef?.current )`) returns `false`
- but the check on line `146` (`if ( anchorRef )`) returns `true`. and therefore `ownerDocument` ends up being `undefined` because `anchorRef` doesn't have an `ownerDocument` property

If you agree, I was planning on opening a separate PR with just this fix once this PR is merged

**Comment by ciampo** [comment]
> FWIW, the Add New Template dropdown probably shouldn't have an arrow. I think it's the only dropdown in the codebase that has one, so that could be seen as an inconsistency.

I'm not an expert when it comes to this part of the codebase, and so I'll happily let other folks make changes to that specific dropdown in a separate PR, if necessary. We can keep this PR focused on the Popover arrow fixes

File: packages/components/src/popover/index.js (hunk 5 above)
Comment (ciampo) [reply packages/components/src/popover/index.js]: > So `14px` is our magic number and I guess it's the one to match our previous design, correct?

Short answer, yes!

Long answer: after a few trials, it looks like, in order to match our previous designs:

- in terms of arrow proportions, the height of the triangle needs to be roughly half of its base (which reflects in the path coordinates in the SVG)
- the previous "arrow size" before the popover refactor was `8px`
- therefore, I factored a height for the triangle of `7px`(plus `1px` border)
- therefore, the base of the triangle is 7 * 2 = `14px`

> Since this number is not affected by anything and is not configurable, is there a reason to have it here and not in the scss file directly?

Good question! When I started working on this PR, most of the arrow styles where applied inline and required some form of math in JS (mostly to get the triangle height, knowing the target triangle base). And so, I decided to use a CSS custom prop to share the value between JS and SCSS.

But now that you point it out, there isn't really a need for it — so I switched to a SCSS variable (which follows more the style of the component) in https://github.com/WordPress/gutenberg/pull/42874/commits/de33f60b3ce48ebcc9edc53bab00f02cab86db7a

File: packages/components/src/popover/style.scss (hunk 2 above)
Comment (ciampo) [reply packages/components/src/popover/style.scss]: Done in https://github.com/WordPress/gutenberg/pull/42874/commits/4a00b8dd36f33c0eff09669629e3ba29237db451

File: packages/components/src/popover/index.js (hunk 1 above)
Comment (ciampo) [reply packages/components/src/popover/index.js]: > The [CSS coding guidelines](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/coding-guidelines.md#css) in this repo are "inspired by BEM" but not exactly BEM, so we want the modifier classes to be named `.is-top`, `.is-right`, etc.

Done in https://github.com/WordPress/gutenberg/pull/42874/commits/097510867480c0187c5e536d16a0a0f11eecb731


> Super nit: why not just use string interpolation here?

I'm not sure I understand your suggestion, can you elaborate a bit more?

**Comment by ciampo** [comment]
All feedback should be addressed. I've also rebased and removed the temporary code for the Storybook example and the `ownerDocument` [fix](https://github.com/WordPress/gutenberg/pull/42874#discussion_r935239671) (which I'll open separate PRs for).

If no one has any objections, I'm planning on merging this PR later today.

File: packages/components/src/popover/index.js (hunk 4 above)
Comment (ciampo) [reply packages/components/src/popover/index.js]: https://github.com/WordPress/gutenberg/pull/42886

File: packages/components/src/popover/index.js (hunk 1 above)
Comment (ntsekouras) [reply packages/components/src/popover/index.js]: I was suggesting:
```
className={`components-popover__arrow is-${ placementData.split( '-' )[ 0 ] }`}
```

Comment (ciampo) [reply packages/components/src/popover/index.js]: I see, thank you for the eplanation!

I personally tend to use the `[ ...classnames ].join(' ')` syntax out of habit — I find it tidier to read, and it allows me to easily add conditional classnames + filter to exclude them if needed (ie.. `[ ...classnames ].filter(Boolean).join(' ')`)
